src/modules/kvs/kvs.c

```diff
@@ -864,6 +866,33 @@ static void kvstxn_wait_error_cb (wait_t *w, int errnum, void *arg)
     kvstxn_set_aux_errnum (kt, errnum);
 }
 
+/* ccan list doesn't appear to have a macro to check if a node exists
+ * on a list */
+static inline bool root_on_work_queue (struct list_node *n)
+{
+    return !(n->next == n->prev && n->next == n);
+}
+
+static void work_queue_append (struct kvs_ctx *ctx,
+                               struct kvsroot *root)
+{
+    if (!root_on_work_queue (&root->work_queue_node))
+        list_add_tail (&ctx->work_queue, &root->work_queue_node);
+}
+
+static void work_queue_remove (struct kvsroot *root)
+{
+    if (root_on_work_queue (&root->work_queue_node))
```

Comment:
I think the test is not needed here.  `list_del_init()` should be safe to call with node on or off list.